### PR TITLE
Fix: Lower C# runtime version to 7.3

### LINKED ISSUE #1873: Dafny 3.4.2 Requires C# 8
While testing Dafny 3.4.2 on our library, which targets C#, we encountered the following error when testing in C#:

```
src/redacted/redacted/redacted/redacted/Source/obj/Debug/netstandard2.0/GeneratedFromDafny.cs(11926,20): error CS8370: Feature 'recursive patterns' is not available in C# 7.3. Please use language version 8.0 or greater.
```

There is no call out in the release notes for 3.4.2 that C#8 will now be required. 

Our library works fine on Dafny 3.4.1. 

Thank you!

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR enforces the runtime c# version to be 7.3 for now.

Fixes #1873

I added an explicit versioning for the c# language in the DafnyRuntime.csproj, which detected advanced features and led me to revert those to usual features. That way, we can stay backward compatible longer.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 - feat: The keyword syntax for functions will change in Dafny version 4. The new command-line option `/functionSyntax` (see `/help`) allows early adoption of the new syntax. (https://github.com/dafny-lang/dafny/pull/1832)
 - fix: No warning "File contains no code" if a file only contains a submodule (https://github.com/dafny-lang/dafny/pull/1840)
 - fix: export-reveals of function-by-method now allows the function body to be ghost (https://github.com/dafny-lang/dafny/pull/1855)
+- fix: Dafnyruntime's c# language version explicitly set to 7.1 to stay backward compatible. (https://github.com/dafny-lang/dafny/pull/1877)
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:7]:
```suggestion
- fix: Regain C# 7.1 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
```
I think the rest is not of interest for the notes.

File: Source/DafnyRuntime/DafnyRuntime.csproj [hunk 2]
```diff
@@ -9,6 +9,7 @@
       <PackageVersion>1.1.0</PackageVersion>
       <TargetFramework>net6.0</TargetFramework>
       <OutputPath>..\..\Binaries\</OutputPath>
+      <LangVersion>7.1</LangVersion>
```
Comment (fabiomadge) [inline Source/DafnyRuntime/DafnyRuntime.csproj:12]: The [default](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version) for even the ancient dotnet versions is `7.3`. Should we maybe target that?

**Review by fabiomadge — DISMISSED** [review]

**Comment by fabiomadge** [comment]
Thanks for taking care of that. I would have gotten to it... eventually

File: Source/DafnyRuntime/DafnyRuntime.csproj (hunk 2 above)
Comment (MikaelMayer) [inline Source/DafnyRuntime/DafnyRuntime.csproj:12]:
```suggestion
      <LangVersion>7.3</LangVersion>
```

**Review by fabiomadge — APPROVED** [review]

**Pull request merged**